src/core/src/dimension.cpp

```diff
@@ -107,7 +107,7 @@ bool Dimension::merge(Dimension& dst, const Dimension& d1, const Dimension& d2)
         t->set_as_equal(d1, d2);
     if (d1.m_label == d2.m_label || d2.m_label == 0)
         dst.m_label = d1.m_label;
-    else if (d1.m_label == 0)
+    else
```

Comment:
Suggestion to add generic specification (with unit tests) which can be used as reference.

I've assumed it was  OK 😄 

```cpp
/**
 * \brief Merges two labels.
 * 
 *  | label_a | label_b | result  |  
 *  |---------|---------|---------|
 *  | X       | 0       | X      |
 *  | 0       | X       | X      |
 *  | X       | X       | X      |
 *  | X       | Y       | ?       |
 *  | Y       | X       | ?       |
 *  | 0       | 0       | 0       |
 *   ...
 *
 * \param label_a 
 * \param label_b 
 * \return size_t 
 */
size_t merge_labels(const size_t label_a, const size_t label_b);
```

Reply:
This piece could be decomposed
```
| X       | Y       | ?       |
| Y       | X       | ?       |
 ```

Logic behind this is based on three valid use cases:
1. both dimensions are equal (and != 1)
2. both dimensions are equal (and == 1)
3. one of dimensions is ==1 and the other is !=1

For cases 1 and 2 we can propagate any label through. For case 3 we have to propagate label of dimension which is != 1.

With dynamic dimensions it is not that simple to distinguish one case from another. We should only propagate smth is we are 100% sure this is valid.

Dimension (-1) with label X merged with Dimension (-1) with label Y could propagate any label as one of them could be ==1 and another could be !=1

Reply:
I find new behavior legit! Let's keep it as a baseline.

Reply:
Great, then I'm going to align and update the following `matmul_propagate_labels` test, to expect label "3" for the third dimension. (This is the reason of failing [CI ](https://dev.azure.com/openvinoci/dldt/_build/results?buildId=396712&view=logs&j=982848f2-7f8d-5e57-730a-8df1e5f991ba&t=ab13fbd5-f238-529f-cad5-2a4f89b3a897&l=4012)now)
https://github.com/openvinotoolkit/openvino/blob/2b701580475499f319079e2f415a9344b8869029/src/core/tests/type_prop/matmul.cpp#L478-L502

Reply:
I've added `merge_labels` function, please re-review